### PR TITLE
Implemented embedded_hal Transfer trait for SPIM and removed EasyDMA limitation for UARTE and TWIM

### PR DESCRIPTION
**Pull request by simonsso**
Hi
Here is an implementation for the embedded_hal transfer trait. #10  It is based on the existing DMA setup code but CS is handled outside in embedded_hal and I set up multiple transfers if size is over the limit.

I also wrote some code for #17 not as suggested in the ticket but that could be refactored later.

### CODE REVIEW

File: nrf52-hal-common/src/spim.rs [hunk 1]
```diff
@@ -55,6 +61,92 @@ impl_spim_ext!(
 /// - The over-read character is hardcoded to `0`.
 pub struct Spim<T>(T);
 
+impl<T> Transfer<u8> for Spim<T> where T: SpimExt
+{
+   type Error = Error;
+
+    fn transfer<'w>(&mut self, words: &'w mut [u8]) -> Result<&'w [u8], Error> {
+        // TODO: some targets have a maxcnt whose size is larger
```
Comment (jamesmunns) [inline nrf52-hal-common/src/spim.rs]: Is this comment still true? I think the feature flags that you do on lines 48-51 means this comment can be removed

File: nrf52-hal-common/src/spim.rs [hunk 2]
```diff
@@ -55,6 +61,92 @@ impl_spim_ext!(
...
+        // than a u8, so this length check is overly restrictive
+        // and could be lifted.
+        if words.len() > MAX_SPI_DMA_SIZE as usize {
+            return Err(Error::TxBufferTooLong);
+        }
+
+        let mut offset:u32 = 0;
+        while offset < words.len() as u32 {
+            let datalen = min(MAX_SPI_DMA_SIZE,(words.len() as u32 )- offset);
```
Comment (jamesmunns) [inline nrf52-hal-common/src/spim.rs]:
```suggestion
            let datalen = min(MAX_SPI_DMA_SIZE, (words.len() as u32) - offset);
```

File: nrf52-hal-common/src/spim.rs [hunk 3]
```diff
@@ -55,6 +61,92 @@ impl_spim_ext!(
 /// - The over-read character is hardcoded to `0`.
 pub struct Spim<T>(T);
 
+impl<T> Transfer<u8> for Spim<T> where T: SpimExt
+{
+   type Error = Error;
+
+    fn transfer<'w>(&mut self, words: &'w mut [u8]) -> Result<&'w [u8], Error> {
```
Comment (jamesmunns) [inline nrf52-hal-common/src/spim.rs]: There is a fair amount of duplicated code here between `Spim<T>::write()` and this method. Would it make sense to refactor some of the common parts into an inline-able function?

File: nrf52-hal-common/src/spim.rs (hunk 1 above)
Comment (simonsso) [reply nrf52-hal-common/src/spim.rs]: Thanks, not only the comment but also the code it commented should be removed!

File: nrf52-hal-common/src/spim.rs (hunk 3 above)
Comment (jamesmunns) [reply nrf52-hal-common/src/spim.rs]: Another option could be to have the `write` function call the `transfer` function, basically only handling the CS GPIO?

Comment (simonsso) [reply nrf52-hal-common/src/spim.rs]: I will see if I could make a nice refactoring of it, almost the same code is on three places.

**Comment by simonsso** [comment]
So I have joined all to one helper function and use it from all three places. But I have not been able to verify the other functions. I don't think moving the handling of CS a few instructions will matter

**Comment by simonsso** [comment]
Some more updates, I applied the EasyDMA fix for UARTE too, and created a shared definition. The same code will probably work for TWIM too but I don't have any thing here to verify that with so I leave it as it is for now.

File: nrf52-hal-common/src/lib.rs [hunk 4]
```diff
@@ -30,8 +30,15 @@ pub mod prelude {
     pub use crate::twim::TwimExt;
     pub use crate::uarte::UarteExt;
 }
-
-
+/// Length of Nordoc EasyDMA differs for MCUs
```
Comment (Yatekii) [inline nrf52-hal-common/src/lib.rs]: Typo ;)

**Comment by simonsso** [comment]
Should I update something more on this PR?

**Comment by jamesmunns** [comment]
Hey @simonsso, sorry for the delay here, I've been working on some other stuff, and haven't been able to take a closer look at the interface you've added here. If @hannobraun has some time, he might be able to provide some suggestions.

**Comment by simonsso** [comment]
SPI is now in new PR #48  -- when the code is reviewed there for SPI the same structure can be applied to TWIM and UART

**Comment by simonsso** [comment]
@jamesmunns  is it ok to make a overwrite-merge (force push to branch) to this PR for the final part, or will it be better if I create a new and abandon this?

**Comment by hannobraun** [comment]
> is it ok to make a overwrite-merge (force push to branch) to this PR for the final part, or will it be better if I create a new and abandon this?

I can't speak for James, but I think it's probably better to open a new PR with the remaining changes. I don't feel strongly about this though, so do what you want :-)

**Comment by jamesmunns** [comment]
@simonsso can this be closed now that #48 has landed?

**Comment by simonsso** [comment]
There are changes on this for i2c and uart too I think but we are close to clean up time for this.

**Comment by simonsso** [comment]
Original change is merged in #48  -- closing this and creating a new for the other changes

**Pull request closed without merging**